content/blog/2019-11-dtplyr-1-0-0/index.Rmarkdown

```diff
@@ -0,0 +1,228 @@
+library(dtplyr)
+library(dplyr, warn.conflicts = FALSE)
+```
+
+Then use `lazy_dt()` to create a "lazy" data table that tracks the operations performed on it:
+
+```{r}
+mtcars2 <- lazy_dt(mtcars)
+```
+
+You can use dplyr verbs with this object as if it was a data frame. But there's a big difference behind the scenes: instead of immediately performing the operation, dtplyr just records what you're doing so when needed it can generate a single, efficient, data.table statement. You can preview the transformation (including the generated data.table code) by printing the result:
+
+```{r}
+mtcars2 %>% 
+  filter(wt < 5) %>% 
+  mutate(l100k = 235.21 / mpg) %>% # liters / 100 km
+  group_by(cyl) %>% 
+  summarise(l100k = mean(l100k))
+```
+
+Generally, however, you should reserve this for exploration and debugging, and instead use `as.data.table()`, `as.data.frame()`, or `as_tibble()` to indicate that you're done writing the transformation and want to access the results:
+
+```{r}
+mtcars2 %>% 
+  filter(wt < 5) %>% 
+  mutate(l100k = 235.21 / mpg) %>% # liters / 100 km
+  group_by(cyl) %>% 
+  summarise(l100k = mean(l100k)) %>% 
+  as_tibble()
+```
+
+## Performance
+
+You might wonder how fast dtplyr is. That's a difficult question to answer in general (because there are so many different ways to use it), so here I'll focus on why dtplyr might be slower than data.table. There are three main reasons: translation cost, copies, and interface mismatch
+
+### Translation cost
+
+Each dplyr verb must do some work to convert dplyr syntax to data.table syntax. We can use the [bench](http://bench.r-lib.org/) package to time the cost of four step pipeline that I used above:
+  
+```{r}
+bench::mark(
+  mtcars2 %>% 
+    filter(wt < 5) %>% 
+    mutate(l100k = 235.21 / mpg) %>% # liters / 100 km
+    group_by(cyl) %>% 
+    summarise(l100k = mean(l100k))
+)
+```
+
+Because this pipeline does not use `as.data.table()` or `print()` it only generates the data.table code, it doesn't run it, so we're timing the translation cost. The translation cost scales with the complexity of the pipeline, not the size of the data, so these timings will apply regardless of the size of the underlying data.
+
+My intial experiments suggest that the translation cost is typically a few milliseconds. Since the computational cost increases with the size of the data, the translation cost becomes a smaller proportion of the total as the data size grows, suggesting the dtplyr does not impose a significant overhead on top of data.table.
+
+Take the following example, which uses large nycflights13 dataset. This isn't really big enough for data.table to really shine, but it's about as big as you can get in an R package. Here I'm going to compute the average arrival delay by destination. It takes raw dplyr about 60ms to do the work. Again, the dtplyr translation is fast, around 1ms, and then computation using data.table only takes about 30ms, almost twice as fast as dplyr.
```

Comment:
```suggestion
Take the following example, which uses the large nycflights13 dataset. This isn't really big enough for data.table to really shine, but it's about as big as you can get in an R package. Here I'm going to compute the average arrival delay by destination. It takes raw dplyr about 60ms to do the work. Again, the dtplyr translation is fast, around 1ms, and then computation using data.table only takes about 30ms, almost twice as fast as dplyr.
```

Reply:
Or I guess it could be **a** large nycflights13 dataset—you just need an article.